cf-oscore/src/main/java/org/eclipse/californium/oscore/ObjectSecurityLayer.java

```diff
@@ -349,4 +382,105 @@ private static boolean shouldProtectRequest(Request request) {
 	private static boolean isProtected(Message message) {
 		return message.getOptions().getOscore() != null;
 	}
+
+	/**
+	 * Check if a message being sent exceeds the MAX_UNFRAGMENTED_SIZE and is
+	 * not using inner block-wise. If so it should not be sent.
+	 * 
+	 * @param message the CoAP message
+	 * @param maxUnfragmentedSize the MAX_UNFRAGMENTED_SIZE value
+	 * 
+	 * @return if the message exceeds the MAX_UNFRAGMENTED_SIZE
+	 */
+	private boolean outgoingExceedsMaxUnfragSize(Message message, boolean outerBlockwise,
+			int maxUnfragmentedSize) {
+
+		boolean usesInnerBlockwise = (message.getOptions().hasBlock1() == true
+				|| message.getOptions().hasBlock2() == true) && outerBlockwise == false;
+
+		if (message.getPayloadSize() > maxUnfragmentedSize && usesInnerBlockwise == false) {
+			return true;
+		} else {
+			return false;
+		}
+
+	}
+
+	/**
+	 * Check if the cumulative payload size for an incoming block-wise transfer
+	 * exceeds the MAX_UNFRAGMENTED_SIZE and is not using inner block-wise. If
+	 * so reception should be aborted. This method is only called when using
+	 * outer block-wise so no need to check that explicitly.
+	 * 
+	 * @param message the CoAP message
+	 * @param exchange the exchange
+	 * @param ctxDb the context database used
+	 * 
+	 * @return if the cumulative payload exceeds the MAX_UNFRAGMENTED_SIZE
+	 */
+	private boolean incomingExceedsMaxUnfragSize(Message message, Exchange exchange, OSCoreCtxDB ctxDb) {
+
+		int cumulativeSize;
+		OSCoreCtx ctx = null;
+		if (message instanceof Request) {
+			ctx = ctxDb.getContext(getRid(message.getOptions().getOscore()));
+			cumulativeSize = blockwiseTransferInfo.getReceivedBodySize(exchange, (Request) message);
+			cumulativeSize += message.getPayloadSize(); // Count current payload
+		} else {
+			ctx = ctxDb.getContextByToken(message.getToken());
+			cumulativeSize = blockwiseTransferInfo.getReceivedBodySize(exchange, (Response) message);
+			cumulativeSize += message.getPayloadSize(); // Count current payload
+		}
+
+		// Do not consider it exceeded if no context is found, it will be
+		// handled elsewhere
+		if (ctx == null) {
+			return false;
+		}
+
+		if (cumulativeSize > ctx.getMaxUnfragmentedSize()) {
```

Comment:
The max size seems to be defined by the rid or token.

On the client side, the uri ctx is transferred to the token context and so taken in account for responses.
I'm not sure, if adding such a "max response size" to a request would not be easier and also protects other transfers, not only oscore.

But the server side, I'm not sure, how it works.
Is that "rid" request specific (it's calculated from the received oscore option)?
Does the server need to be prepared for that exact request?
Or is that "rid" more resource specific?

Reply:
> The max size seems to be defined by the rid or token.

The max size is defined in the OSCORE context used for a particular response or request. And the context used (for incoming messages) is retrieved by using either the Recipient ID (RID) from the OSCORE option of an incoming request, or the Token for an incoming response.

A user or application developer can set specific size limits per OSCORE context (perhaps one server needs a lower limit so the context shared with that server can be configured to have a lower limit).

> But the server side, I'm not sure, how it works.
> Is that "rid" request specific (it's calculated from the received oscore option)?
> Does the server need to be prepared for that exact request?
> Or is that "rid" more resource specific?

Yes indeed, the Recipient ID (RID) is request specific. It is contained in the OSCORE option and parsed from there. By doing that the server can identity the OSCORE context associated with the incoming request and retrieve the particular size limit from that context.

The server does not need to be prepared for that particular request except for having an OSCORE context associated to the RID in the OSCORE option of the request. The RID is not resource-specific.